### PR TITLE
Feature: Banning Users

### PR DESCRIPTION
**Pull request by davidnsai**
I completed the ban feature, it can be accessed using auth()->user()->ban(); the ban method takes in an optional string which is saved on the database as the reason for the ban, this is then displayed to the user when they try to log in. I did not use status and status_message because I felt they were not descriptive enough. An unBan() method was also added to allow access to a user again.

### CODE REVIEW

**Comment by jozefrebjak** [comment]
@davidnsai Please add new migration with altering table when you added new columns.

**Comment by davidnsai** [comment]
> @davidnsai Please add new migration with altering table when you added new columns.

@jozefrebjak I just added a new migration file altering the table

File: src/Authorization/Traits/Authorizable.php [hunk 1]
```diff
@@ -405,4 +430,13 @@ private function getConfigPermissions(): array
             ? array_keys(setting('AuthGroups.permissions'))
             : array_keys(config('AuthGroups')->permissions);
     }
+
+    /**
+     * modifies the banned status of the user
+     */
+    private function modifyBanStatus(): void
+    {
+        $model = model(UserModel::class);
```
Comment (lonnieezell) [inline src/Authorization/Traits/Authorizable.php:439]: I believe you can get the User Provider (which might be a class that extends the UserModel) directly from the `auth()` helper. Double-check me on that, though. 

```suggestion
        $model = auth()->getProvider();
```

File: src/Database/Migrations/2022-12-13-200557_add_ban_columns.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+            'ban_message' => ['type' => 'varchar', 'after' => 'banned', 'constraint' => 255, 'null' => true],
+        ];
+        
+        $this->forge->addColumn('users', $fields);
+    }
+
+    // --------------------------------------------------------------------
+
+    public function down(): void
+    {
```
Comment (lonnieezell) [inline src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:25]: You should drop these new columns in the `down()` method.

File: src/Database/Migrations/2022-12-13-200557_add_ban_columns.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+{
+    public function up(): void
+    {
+        // Users Table
+        $fields = [
+            'banned'      => ['type' => 'tinyint', 'after' => 'active', 'constraint' => 1, 'null' => 0, 'default' => 0],
+            'ban_message' => ['type' => 'varchar', 'after' => 'banned', 'constraint' => 255, 'null' => true],
+        ];
+        
+        $this->forge->addColumn('users', $fields);
```
Comment (lonnieezell) [inline src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:19]: I think we can get rid of the `status` and `status_message` columns now. We will need to ensure we let people know about that in the changelog, though.

**Review by lonnieezell — CHANGES_REQUESTED** [review]
Looks pretty good, thanks! A few small things to take care of, though. 

Need to add docs about it in [authentication.md](https://github.com/codeigniter4/shield/blob/develop/docs/authentication.md). 

It doesn't look like we have one, yet, but we need to make a note of dropping the status and status_message columns in a new `CHANGELOG.md` file.

**Comment by kenjis** [comment]
This error message does not help us.

```
PHP Fatal error:  Uncaught _PHPStan_9a6ded56a\Nette\Schema\ValidationException: Unexpected item 'parameters › deprecationRulesInstalled'. in phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/schema/src/Schema/Processor.php:75

Stack trace:
Fatal error: Uncaught _PHPStan_9a6ded56a\Nette\Schema\ValidationException: Unexpected item 'parameters › deprecationRulesInstalled'. in phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/schema/src/Schema/Processor.php:75
#0 phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/schema/src/Schema/Processor.php(38): _PHPStan_9a6ded56a\Nette\Schema\Processor->throwsErrors()
Stack trace:
#0 phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/schema/src/Schema/Processor.php(38): _PHPStan_9a6ded56a\Nette\Schema\Processor->throwsErrors()
#1 phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/DependencyInjection/ParametersSchemaExtension.php(43): _PHPStan_9a6ded56a\Nette\Schema\Processor->process()
#2 phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/di/src/DI/Compiler.php(182): PHPStan\DependencyInjection\ParametersSchemaExtension->loadConfiguration()
#3 phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/di/src/DI/Compiler.php(158): _PHPStan_9 in phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/schema/src/Schema/Processor.php on line 75
#1 phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/DependencyInjection/ParametersSchemaExtension.php(43): _PHPStan_9a6ded56a\Nette\Schema\Processor->process()
#2 phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/di/src/DI/Compiler.php(182): PHPStan\DependencyInjection\ParametersSchemaExtension->loadConfiguration()
#3 phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/di/src/DI/Compiler.php(158): _PHPStan_9 in phar:///home/runner/.composer/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/schema/src/Schema/Processor.php on line 75
Error: Process completed with exit code 255.
```
https://github.com/codeigniter4/shield/actions/runs/3689118746/jobs/6244710242

File: src/Authorization/Traits/Authorizable.php [hunk 4]
```diff
@@ -292,6 +293,30 @@ public function inGroup(string ...$groups): bool
...
+    {
+        $this->banned      = '1';
+        $this->ban_message = $reason;
+        $this->modifyBanStatus();
+
+        return $this;
+    }
+
+    /**
+     * Bans the user from logging in.
```
Comment (sammyskills) [inline src/Authorization/Traits/Authorizable.php:309]: I believe this commit should relate to what the function does.
```
Unban a user to permit logging in.
```

**Review by sammyskills — CHANGES_REQUESTED** [review]

**Comment by kenjis** [comment]
Please fix the errors in tests.
```
There were 2 errors:

1) Tests\Authentication\Filters\SessionFilterTest::testFilterSuccess
TypeError: CodeIgniter\Shield\Authentication\Authenticators\Session::isBanned(): Return value must be of type bool, string returned

/home/runner/work/shield/shield/src/Authentication/Authenticators/Session.php:538
/home/runner/work/shield/shield/src/Filters/SessionAuth.php:48
/home/runner/work/shield/shield/vendor/codeigniter4/framework/system/Filters/Filters.php:173
/home/runner/work/shield/shield/vendor/codeigniter4/framework/system/CodeIgniter.php:455
/home/runner/work/shield/shield/vendor/codeigniter4/framework/system/CodeIgniter.php:351
/home/runner/work/shield/shield/vendor/codeigniter4/framework/system/Test/FeatureTestTrait.php:187
/home/runner/work/shield/shield/vendor/codeigniter4/framework/system/Test/FeatureTestTrait.php:219
/home/runner/work/shield/shield/tests/Authentication/Filters/SessionFilterTest.php:38

2) Tests\Authentication\Filters\SessionFilterTest::testRecordActiveDate
TypeError: CodeIgniter\Shield\Authentication\Authenticators\Session::isBanned(): Return value must be of type bool, string returned

/home/runner/work/shield/shield/src/Authentication/Authenticators/Session.php:538
/home/runner/work/shield/shield/src/Filters/SessionAuth.php:48
/home/runner/work/shield/shield/vendor/codeigniter4/framework/system/Filters/Filters.php:173
/home/runner/work/shield/shield/vendor/codeigniter4/framework/system/CodeIgniter.php:455
/home/runner/work/shield/shield/vendor/codeigniter4/framework/system/CodeIgniter.php:351
/home/runner/work/shield/shield/vendor/codeigniter4/framework/system/Test/FeatureTestTrait.php:187
/home/runner/work/shield/shield/vendor/codeigniter4/framework/system/Test/FeatureTestTrait.php:219
/home/runner/work/shield/shield/tests/Authentication/Filters/SessionFilterTest.php:55

ERRORS!
Tests: 217, Assertions: 551, Errors: 2.
```
https://github.com/codeigniter4/shield/actions/runs/3693320215/jobs/6253174013

**Comment by davidnsai** [comment]
> Looks pretty good, thanks! A few small things to take care of, though.
> 
> Need to add docs about it in [authentication.md](https://github.com/codeigniter4/shield/blob/develop/docs/authentication.md).
> 
> It doesn't look like we have one, yet, but we need to make a note of dropping the status and status_message columns in a new `CHANGELOG.md` file.

I was thinking it would be better to add the  docs in the [authorization.md](https://github.com/codeigniter4/shield/blob/develop/docs/authentication.md) as the feature does not deal with authenticating the user but deals with authorizing them to access the protected parts of websites

**Comment by lonnieezell** [comment]
> I was thinking it would be better to add the docs in the [authorization.md](https://github.com/codeigniter4/shield/blob/develop/docs/authentication.md) as the feature does not deal with authenticating the user but deals with authorizing them to access the protected parts of websites

Doh. You're correct.

**Comment by lonnieezell** [comment]
@davidnsai did you mean to close this PR? Was there an issue?

**Comment by davidnsai** [comment]
@lonnieezell I didn't mean to close it. It must have happened when I closed one which I was going to make later on. So I realized that with the access tokens, the filter handling that redirects to login if the access token is not present or is invalid I felt this is not the best implementation of this because you want to send a RESTful message back to the client so unlike having a redirect, I felt it would be better if a json response with status code 403 is sent back to the client that way a message can even be appended into the body of the response. But I felt this can be done in a later PR after this is resolved. I will reopen it

**Comment by datamweb** [comment]
> So I realized that with the access tokens, the filter handling that redirects to login if the access token is not present or is invalid I felt this is not the best implementation of this because you want to send a RESTful message back to the client so unlike having a redirect,

@lonnieezell I have never been able to understand why we have something like this in filter:
https://github.com/codeigniter4/shield/blob/3fede093f21cca09e7997fc46cf8c81e0b9a648d/src/Filters/TokenAuth.php#L51-L53
Can you explain to me what is the need for the **login page** to return in **Restful**.
Please see here.#524 and #433

**Comment by lonnieezell** [comment]
The redirect to login in the TokenAuth class looks to be a mistake and should be handled as @davidnsai mentioned above.

**Comment by davidnsai** [comment]
@lonnieezell  I'll include that in my next PR

**Comment by lonnieezell** [comment]
@davidnsai no need. I've got a PR that should get submitted tonight dealing with active status, and I've fixed it in there.

**Comment by davidnsai** [comment]
Also, I have something else I feel should be resolved. So I realised that we don't have a configuration for selecting a database group on shield I feel it would be a good feature to implement in the event that a distributed application has to be built.

**Pull request closed without merging**